flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/SortUtil.scala

```diff
@@ -0,0 +1,367 @@
+ */
+
+trait UntypedOrdering extends Serializable{
+  def compare(x: Any, y: Any): Int
+
+}
+
+class LongOrdering(implicit ord: Ordering[Long]) extends UntypedOrdering {
+
+  override def compare(x: Any, y: Any): Int = {
+    val xL = x.asInstanceOf[Long]
+    val yL = y.asInstanceOf[Long]
+    ord.compare(xL, yL)
+  }
+}
+
+class IntOrdering(implicit ord: Ordering[Int]) extends UntypedOrdering {
+
+  override def compare(x: Any, y: Any): Int = {
+    val xI = x.asInstanceOf[Int]
+    val yI = y.asInstanceOf[Int]
+    ord.compare(xI, yI)
+  }
+}
+
+class FloatOrdering(implicit ord: Ordering[Float]) extends UntypedOrdering {
+
+  override def compare(x: Any, y: Any): Int = {
+    val xF = x.asInstanceOf[Float]
+    val yF = y.asInstanceOf[Float]
+    ord.compare(xF, yF)
+  }
+}
+
+class DoubleOrdering(implicit ord: Ordering[Double]) extends UntypedOrdering {
+
+  override def compare(x: Any, y: Any): Int = {
+    val xD = x.asInstanceOf[Double]
+    val yD = y.asInstanceOf[Double]
+    ord.compare(xD, yD)
+  }
+}
+
+class DecimalOrdering(implicit ord: Ordering[BigDecimal]) extends UntypedOrdering {
+
+  override def compare(x: Any, y: Any): Int = {
+    val xBD = x.asInstanceOf[BigDecimal]
+    val yBD = y.asInstanceOf[BigDecimal]
+    ord.compare(xBD, yBD)
+  }
+}
+
+class ByteOrdering(implicit ord: Ordering[Byte]) extends UntypedOrdering {
+
+  override def compare(x: Any, y: Any): Int = {
+    val xB = x.asInstanceOf[Byte]
+    val yB = y.asInstanceOf[Byte]
+    ord.compare(xB, yB)
+  }
+}
+
+class SmallOrdering(implicit ord: Ordering[Short]) extends UntypedOrdering {
+
+  override def compare(x: Any, y: Any): Int = {
+    val xS = x.asInstanceOf[Short]
+    val yS = y.asInstanceOf[Short]
+    ord.compare(xS, yS)
+  }
+}
+
+class StringOrdering(implicit ord: Ordering[String]) extends UntypedOrdering {
+
+  override def compare(x: Any, y: Any): Int = {
+    val xS = x.asInstanceOf[String]
+    val yS = y.asInstanceOf[String]
+    ord.compare(xS, yS)
+  }
+}
+
+/**
+ * Ordering object for timestamps. As there is no implicit Ordering for [java.sql.Timestamp]
+ * we need to compare based on the Long value of the timestamp
+ */
+class TimestampOrdering(implicit ord: Ordering[Long]) extends UntypedOrdering {
+
+  override def compare(x: Any, y: Any): Int = {
+    val xTs = x.asInstanceOf[Timestamp]
+    val yTs = y.asInstanceOf[Timestamp]
+    ord.compare(xTs.getTime, yTs.getTime)
+  }
+}
+
+
+
+/**
+ * Called every time when a multi-output aggregation result should be materialized.
+ * The returned values could be either an early and incomplete result
+ * (periodically emitted as data arrive) or the final result of the
+ * aggregation.
+ *
+ * @param accumulator the accumulator which contains the current
+ *                    aggregated results
+ * @return the aggregation result
+ */
+abstract class MultiOutputAggregateFunction[T] extends AggregateFunction[T] {
```

Comment:
I think this is debatable...i can see also aggregates that would just have more values (e.g., aggregate last 10 values).
Nevertheless, this is irrelevant if we move to the implementation in the processing function